Mathlib/Combinatorics/Flow/Basic.lean

```diff
@@ -0,0 +1,232 @@
+-/
+
+universe u_V u_R
+
+open BigOperators
+
+section NetworkDefinition
+
+variable (V : Type u_V) (R : Type u_R) [Zero R] [LE R]
+
+/-- A flow network assigning non-negative capacities to all pairs of vertices. -/
+structure Network extends Digraph V where
+  Adj _ _ := True
+  /-- The assigned capacities. -/
+  cap : V → V → R
+  nonneg : ∀ u v, 0 ≤ cap u v
+  cap_zero_of_not_adj : ∀ u v, ¬Adj u v → cap u v = 0 := by intro u v; aesop
+
+/-- A `Network` where all capacities are symmetrical. -/
+structure UndirectedNetwork extends Network V R where
+  symm : ∀ u v, cap u v = cap v u
+
+end NetworkDefinition
+
+variable {V : Type u_V} {R : Type u_R}
+
+namespace Network
+
+section PseudoFlowDefinition
+
+variable [Zero R] [Preorder R]
+
+/-- A `PseudoFlow` is an assignment of non-negative flow values to each arc of the network, such
+that the assigned amount does not exceed the capacity.
+
+Note that this definition, in contrast to what can sometimes be found in the literature, does
+not require that $f(u, v) = -f(v, u)$. This extra requirement eliminates cycles between two
+vertices and makes some theorem statements shorter. However, since larger cycles are still
+permitted, this usually does not help in proofs. To the contrary, defining a flow becomes more
+cumbersome, sometimes significantly so. -/
+@[ext]
+structure PseudoFlow (N : Network V R) where
+  /-- The assignment of flow values to each arc.
+  Do NOT use. Use the coercion to function instead. -/
+  toFun : V → V → R
+  nonneg' : ∀ u v, 0 ≤ toFun u v
+  capacity' : ∀ u v, toFun u v ≤ N.cap u v
+
+namespace PseudoFlow
+
+variable {N : Network V R}
+
+instance instFunLike : FunLike N.PseudoFlow V (V → R) where
+  coe := PseudoFlow.toFun
+  coe_injective' f g h := by cases f; cases g; congr
+
+variable (f : N.PseudoFlow)
+
+lemma nonneg : ∀ u v, 0 ≤ f u v := f.nonneg'
+
+lemma capacity : ∀ u v, f u v ≤ N.cap u v := f.capacity'
+
+instance : Zero N.PseudoFlow where
+  zero.toFun _ _ := 0
+  zero.nonneg' _ _ := le_rfl
+  zero.capacity' := N.nonneg
+
+end PseudoFlow
+end PseudoFlowDefinition
+
+variable [Fintype V]
+
+namespace PseudoFlow
+
+variable [Preorder R] [AddCommMonoid R] {N : Network V R} (f : N.PseudoFlow) (v : V)
+
+/-- The incoming flow of a vertex. --/
+def incoming := ∑ u, f u v
+/-- The outgoing flow of a vertex. --/
+def outgoing := ∑ w, f v w
+/-- The excess flow of a vertex (the incoming flow minus the outgoing flow). --/
+def excess [Sub R] := f.incoming v - f.outgoing v
+
+end PseudoFlow
+
+section PreFlowDefinition
+
+variable [Preorder R] [AddCommMonoid R] (N : Network V R) (s t : V)
+
+/-- A `PreFlow` from `s` to `t` is a `PseudoFlow` such that all vertices except `s` and `t` have
+non-negative excess, that is, the outgoing flow is at most as much as the incoming flow. -/
+@[ext]
+structure PreFlow extends N.PseudoFlow where
+  excess_nonneg : ∀ v, v ≠ s ∧ v ≠ t → toPseudoFlow.outgoing v ≤ toPseudoFlow.incoming v
```

Comment:
```suggestion
  excess_nonneg : ∀ v, v ≠ s → v ≠ t → toPseudoFlow.outgoing v ≤ toPseudoFlow.incoming v
```
The usual Lean pattern is to curry constructors in this way. You'll need to change a couple of proofs for this, but it should be easy to adjust to!